# US Monster Jobs Dataset Cleansing

Personal checklist:
-> Standarization of data: Fix inconsistent column names and convert into a standard format

In [2]:
import pandas as pd

import numpy as np

In [3]:
df = pd.read_csv('C:\\Users\\jorge\\Desktop\\monster_com-job_sample.csv')

# Overview of dataset

In [4]:
df.shape

(22000, 14)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22000 entries, 0 to 21999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   country          22000 non-null  object
 1   country_code     22000 non-null  object
 2   date_added       122 non-null    object
 3   has_expired      22000 non-null  object
 4   job_board        22000 non-null  object
 5   job_description  22000 non-null  object
 6   job_title        22000 non-null  object
 7   job_type         20372 non-null  object
 8   location         22000 non-null  object
 9   organization     15133 non-null  object
 10  page_url         22000 non-null  object
 11  salary           3446 non-null   object
 12  sector           16806 non-null  object
 13  uniq_id          22000 non-null  object
dtypes: object(14)
memory usage: 2.3+ MB


# 1. Standarization of data
-> Column names/format: I don't see any relevant errors in the column name format. Am I missing something?
 <br />No caps, no format differences between one name and another.

In [6]:
df.columns

Index(['country', 'country_code', 'date_added', 'has_expired', 'job_board',
       'job_description', 'job_title', 'job_type', 'location', 'organization',
       'page_url', 'salary', 'sector', 'uniq_id'],
      dtype='object')

# 2. Data Type Conversion
Data in df is of 'object' type. We'll have to convert the data type of the columns according to the data in each one of them.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22000 entries, 0 to 21999
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   country          22000 non-null  object
 1   country_code     22000 non-null  object
 2   date_added       122 non-null    object
 3   has_expired      22000 non-null  object
 4   job_board        22000 non-null  object
 5   job_description  22000 non-null  object
 6   job_title        22000 non-null  object
 7   job_type         20372 non-null  object
 8   location         22000 non-null  object
 9   organization     15133 non-null  object
 10  page_url         22000 non-null  object
 11  salary           3446 non-null   object
 12  sector           16806 non-null  object
 13  uniq_id          22000 non-null  object
dtypes: object(14)
memory usage: 2.3+ MB


# 2.1 Empty cells per column?

We have a ton of missing data in date_added, job_type, organization, salary and sector.

In [8]:
df.isnull().sum()

country                0
country_code           0
date_added         21878
has_expired            0
job_board              0
job_description        0
job_title              0
job_type            1628
location               0
organization        6867
page_url               0
salary             18554
sector              5194
uniq_id                0
dtype: int64

The only value in the 'Country' column is the US. I guess it makes sense when using a 'US Monster Jobs' dataset.

In [9]:
print(df['country'].value_counts())

print('-----------------------------------')

print(df['country_code'].value_counts())

United States of America    22000
Name: country, dtype: int64
-----------------------------------
US    22000
Name: country_code, dtype: int64


# 2.2 Convert date_added to date dtype

In [10]:
df.loc[df['date_added'].notnull()]

country country_code date_added has_expired  \
133    United States of America           US  5/10/2016          No   
140    United States of America           US  5/13/2016          No   
251    United States of America           US   5/9/2016          No   
279    United States of America           US  6/10/2016          No   
366    United States of America           US   1/2/2017          No   
...                         ...          ...        ...         ...   
20760  United States of America           US  9/27/2016          No   
21342  United States of America           US  3/30/2016          No   
21391  United States of America           US  3/24/2016          No   
21631  United States of America           US   4/4/2016          No   
21696  United States of America           US  3/24/2016          No   

              job_board                                    job_description  \
133    jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   
140    jobs.monster.com  Equal Opportunity Employer: Minority/Female/Di...   
251    jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   
279    jobs.monster.com  At American Family Insurance, we're firmly com...   
366    jobs.monster.com  Description The Opportunity The Vehicle Mainte...   
...                 ...                                                ...   
20760  jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   
21342  jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   
21391  jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   
21631  jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   
21696  jobs.monster.com  #TrackingJobBody table, #TrackingJobBody a {<b...   

                                               job_title            job_type  \
133                 Multibed Technician Job in Deer Park  Full Time Employee   
140     Principal Cyber Security Engineer Job in Houston  Full Time Employee   
251                 Field Supervisor IS Job in Deer Park  Full Time Employee   
279    Insurance Sales - Customer Service Job in Eden...  Full Time Employee   
366             Vehicle Maintenance Mechanic - Las Vegas  Full Time Employee   
...                                                  ...                 ...   
20760         Central Maintenance Planner Job in Norwell  Full Time Employee   
21342                   Branch Manager Job in Cincinnati  Full Time Employee   
21391             Field Service Driver Job in Cincinnati  Full Time Employee   
21631            Field Project Manager Job in Cincinnati  Full Time Employee   
21696    Environmental Technician CBPP Job in Cincinnati  Full Time Employee   

                     location                           organization  \
133             Deer Park, TX                   Other/Not Classified   
140               Houston, TX  Computer SoftwareComputer/IT Services   
251             Deer Park, TX                   Other/Not Classified   
279    Eden Prairie, MN 55344                              Insurance   
366             Las Vegas, NV                   Energy and Utilities   
...                       ...                                    ...   
20760             Norwell, MA                   Other/Not Classified   
21342          Cincinnati, OH                   Other/Not Classified   
21391          Cincinnati, OH                   Other/Not Classified   
21631          Cincinnati, OH                   Other/Not Classified   
21696          Cincinnati, OH                   Other/Not Classified   

                                                page_url  \
133    http://jobview.monster.com/Multibed-Technician...   
140    http://jobview.monster.com/Principal-Cyber-Sec...   
251    http://jobview.monster.com/Field-Supervisor-IS...   
279    http://jobview.monster.com/insurance-sales-cus...   
366    http://jobview.monster.com/vehicle-maintenance...   
...                                             

In [11]:
lista = [(i + ''': ''' + str(df.dtypes[i])) for i in dict(df.dtypes)]

print(lista)

['country: object', 'country_code: object', 'date_added: object', 'has_expired: object', 'job_board: object', 'job_description: object', 'job_title: object', 'job_type: object', 'location: object', 'organization: object', 'page_url: object', 'salary: object', 'sector: object', 'uniq_id: object']


# Fix salary column

# Visualize full df rows

Code in next cell allows us to view rows in text editor. 

In [12]:
with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(df[['salary']].head(15))

                           salary
0                             NaN
1                             NaN
2                             NaN
3                             NaN
4                             NaN
5                             NaN
6                             NaN
7                             NaN
8                             NaN
9                             NaN
10                            NaN
11                            NaN
12                            NaN
13           9.00 - 13.00 $ /hour
14  80,000.00 - 95,000.00 $ /year


I wanted to have a better idea of the data in the column.

The "Salary" column is a mess. We got:\
80,000.00 - 95,000.00 /year\
45,000.00 - 100,000.00 /yearBonus, Benefits,\
40,000.00 - 50,000.00 /yearsalary\
56,000.00 - 64,000.00 /yearHighly Competitiv\
13.00 - 16.00 /year\
0.00 - 90,000.00 /year\
45,000.00+ /year\
0.00 - 1.00 /year

Next code block shows rows with "year" string in it. "Salary" column.

In [13]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('year') == True)

with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(df.loc[filt]['salary'].head(15))

14                         80,000.00 - 95,000.00 $ /year
19                         60,000.00 - 72,000.00 $ /year
29                        70,000.00 - 100,000.00 $ /year
32                        75,000.00 - 100,000.00 $ /year
36                         68,000.00 - 72,000.00 $ /year
41                         58,000.00 - 65,000.00 $ /year
61     45,000.00 - 100,000.00 $ /yearBonus, Benefits,...
64                   40,000.00 - 50,000.00 $ /yearsalary
82                         80,000.00 - 90,000.00 $ /year
83                         35,000.00 - 45,000.00 $ /year
88                        80,000.00 - 100,000.00 $ /year
100                        31,000.00 - 33,000.00 $ /year
117                      100,000.00 - 120,000.00 $ /year
127                       75,000.00 - 100,000.00 $ /year
132                                    $50,000.00+ /year
Name: salary, dtype: object


# Add extra columns for analysis

In [14]:
df.columns.get_loc('salary')

11

In [15]:
df.insert(12, 'from_salary', np.nan)

df.columns

Index(['country', 'country_code', 'date_added', 'has_expired', 'job_board',
       'job_description', 'job_title', 'job_type', 'location', 'organization',
       'page_url', 'salary', 'from_salary', 'sector', 'uniq_id'],
      dtype='object')

In [16]:
df.insert(13, 'to_salary', np.nan)

df.columns

Index(['country', 'country_code', 'date_added', 'has_expired', 'job_board',
       'job_description', 'job_title', 'job_type', 'location', 'organization',
       'page_url', 'salary', 'from_salary', 'to_salary', 'sector', 'uniq_id'],
      dtype='object')

In [17]:
df.insert(14, 'yearly_hourly', np.nan)

df.columns

Index(['country', 'country_code', 'date_added', 'has_expired', 'job_board',
       'job_description', 'job_title', 'job_type', 'location', 'organization',
       'page_url', 'salary', 'from_salary', 'to_salary', 'yearly_hourly',
       'sector', 'uniq_id'],
      dtype='object')

In [18]:
df.index.names = ['id']

# Split salary string
Now that I got the columns I'm splitting the strings into each one of them.\
Try to make sense of the data.

In [19]:
df['yearly_hourly']

id
0       NaN
1       NaN
2       NaN
3       NaN
4       NaN
         ..
21995   NaN
21996   NaN
21997   NaN
21998   NaN
21999   NaN
Name: yearly_hourly, Length: 22000, dtype: float64

In [43]:
df['from_salary'] = df['from_salary'].astype(str)

df['to_salary'] = df['to_salary'].astype(str)


df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22000 entries, 0 to 21999
Data columns (total 17 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   country          22000 non-null  object
 1   country_code     22000 non-null  object
 2   date_added       122 non-null    object
 3   has_expired      22000 non-null  object
 4   job_board        22000 non-null  object
 5   job_description  22000 non-null  object
 6   job_title        22000 non-null  object
 7   job_type         20372 non-null  object
 8   location         22000 non-null  object
 9   organization     15133 non-null  object
 10  page_url         22000 non-null  object
 11  salary           3446 non-null   object
 12  from_salary      22000 non-null  object
 13  to_salary        22000 non-null  object
 14  yearly_hourly    1758 non-null   object
 15  sector           16806 non-null  object
 16  uniq_id          22000 non-null  object
dtypes: object(17)
memory usage: 2.9

# Salary/Year Loop

Set "Yearly_Hourly" column to "year" where "Salary" column contains "year"

In [45]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('year') == True)

for i in df.index[filt == True]:
    from_sal = df.iloc[i, df.columns.get_loc('from_salary')]
    to_sal = df.iloc[i, df.columns.get_loc('to_salary')]
    
    df.iloc[i, df.columns.get_loc('yearly_hourly')] = 'year'
    
    split_salary = df.iloc[i, df.columns.get_loc('salary')].split('-', 1)  # Split "Salary" string in 2. That's why we specify "1" inside the split. That means first ocurrence.
    
    if len(split_salary) == 1:
        df.at[i, 'from_salary'] = np.nan   # Set to NaN while I find what to do with salaries without specific range like "Up to $60,000"
        df.at[i, 'to_salary'] = split_salary[0]   # Set to the max value, for ex. if it says "Up to $60,000" then it is $60,000
    elif len(split_salary) == 2:
        df.at[i, 'from_salary'] = split_salary[0]   # We assign FIRST value of the split. Most of the strings are of the "$50,000 - $70,000" type
        df.at[i, 'to_salary'] = split_salary[1]   # We assign SECOND value of the split. Most of the strings are of the "$50,000 - $70,000" type

# Salary/Hour Loop

In [49]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('hour') == True)

for i in df.index[filt == True]:
    from_sal = df.iloc[i, df.columns.get_loc('from_salary')]
    to_sal = df.iloc[i, df.columns.get_loc('to_salary')]
    
    df.iloc[i, df.columns.get_loc('yearly_hourly')] = 'hour'
    
    split_salary = df.iloc[i, df.columns.get_loc('salary')].split('-', 1)
    
    if len(split_salary) == 1:
        df.at[i, 'from_salary'] = np.nan
        df.at[i, 'to_salary'] = split_salary[0]
    elif len(split_salary) == 2:
        df.at[i, 'from_salary'] = split_salary[0]
        df.at[i, 'to_salary'] = split_salary[1]

# Salary per Month

In [78]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains("""/month""") == True)

for i in df.index[filt == True]:
    from_sal = df.iloc[i, df.columns.get_loc('from_salary')]
    to_sal = df.iloc[i, df.columns.get_loc('to_salary')]
    
    df.iloc[i, df.columns.get_loc('yearly_hourly')] = 'month'
    
    split_salary = df.iloc[i, df.columns.get_loc('salary')].split('-', 1)
    
    if len(split_salary) == 1:
        df.at[i, 'from_salary'] = np.nan
        df.at[i, 'to_salary'] = split_salary[0]
    elif len(split_salary) == 2:
        df.at[i, 'from_salary'] = split_salary[0]
        df.at[i, 'to_salary'] = split_salary[1]

# Up to... Loop

In [66]:
filt_upto = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('Up to') == True)

df.loc[filt_upto]['salary']

for i in df.index[filt_upto == True]:
    split_salary = df.iloc[i, df.columns.get_loc('salary')].split('$', 1)
    if len(split_salary[1]) > 5:
        df.at[i, 'from_salary'] = np.nan
        df.at[i, 'to_salary'] = split_salary[1]
        df.at[i, 'yearly_hourly'] = 'upto'
    elif len(split_salary[1]) < 6:
        df.at[i, 'from_salary'] = np.nan
        df.at[i, 'to_salary'] = split_salary[1]
        df.at[i, 'yearly_hourly'] = 'uptohour'

# View year results

In [51]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('year') == True)

df.loc[filt][['salary', 'from_salary', 'to_salary', 'yearly_hourly']]

salary  from_salary  \
id                                                                      
14                         80,000.00 - 95,000.00 $ /year   80,000.00    
19                         60,000.00 - 72,000.00 $ /year   60,000.00    
29                        70,000.00 - 100,000.00 $ /year   70,000.00    
32                        75,000.00 - 100,000.00 $ /year   75,000.00    
36                         68,000.00 - 72,000.00 $ /year   68,000.00    
...                                                  ...          ...   
21982                      70,000.00 - 80,000.00 $ /year   70,000.00    
21987                                  $80,000.00+ /year          NaN   
21995               120,000.00 - 160,000.00 $ /yearbonus  120,000.00    
21996                      45,000.00 - 60,000.00 $ /year   45,000.00    
21999  40,000.00 - 46,000.00 $ /year+ annual bonus (u...   40,000.00    

                                               to_salary yearly_hourly  
id                                                                      
14                                     95,000.00 $ /year          year  
19                                     72,000.00 $ /year          year  
29                                    100,000.00 $ /year          year  
32                                    100,000.00 $ /year          year  
36                                     72,000.00 $ /year          year  
...                                                  ...           ...  
21982                                  80,000.00 $ /year          year  
21987                                  $80,000.00+ /year          year  
21995                            160,000.00 $ /yearbonus          year  
21996                                  60,000.00 $ /year          year  
21999   46,000.00 $ /year+ annual bonus (up to 15% of...          year  

[1758 rows x 4 columns]

# View hour results

In [52]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('hour') == True)

df.loc[filt][['salary', 'from_salary', 'to_salary', 'yearly_hourly']]

salary from_salary  \
id                                                       
13                    9.00 - 13.00 $ /hour       9.00    
30                   62.00 - 81.00 $ /hour      62.00    
43                   15.00 - 16.00 $ /hour      15.00    
68     13.75 - 16.75 $ /hourYear End Bonus      13.75    
80                   40.00 - 50.00 $ /hour      40.00    
...                                    ...         ...   
21909                17.00 - 21.00 $ /hour      17.00    
21912                20.00 - 22.00 $ /hour      20.00    
21913                50.00 - 55.00 $ /hour      50.00    
21927                30.00 - 35.00 $ /hour      30.00    
21998                25.00 - 28.00 $ /hour      25.00    

                          to_salary yearly_hourly  
id                                                 
13                    13.00 $ /hour          hour  
30                    81.00 $ /hour          hour  
43                    16.00 $ /hour          hour  
68      16.75 $ /hourYear End Bonus          hour  
80                    50.00 $ /hour          hour  
...                             ...           ...  
21909                 21.00 $ /hour          hour  
21912                 22.00 $ /hour          hour  
21913                 55.00 $ /hour          hour  
21927                 35.00 $ /hour          hour  
21998                 28.00 $ /hour          hour  

[1145 rows x 4 columns]

# View Monthly

In [79]:
filt = (pd.isna(df['salary']) == False) & (df['salary'].str.contains("""/month""") == True)

df.loc[filt][['salary', 'from_salary', 'to_salary', 'yearly_hourly']]

salary from_salary  \
id                                                             
2662              17,688.94 - 20,971.00 $ /month  17,688.94    
2900                3,137.89 - 3,865.66 $ /month   3,137.89    
3300              17,688.94 - 20,971.00 $ /month  17,688.94    
3317                6,833.33 - 8,060.00 $ /month   6,833.33    
3345                4,224.31 - 6,579.41 $ /month   4,224.31    
3622                4,905.22 - 6,579.41 $ /month   4,905.22    
5949                4,362.39 - 5,445.57 $ /month   4,362.39    
12333               1,800.00 - 3,500.00 $ /month   1,800.00    
13554               5,882.93 - 7,883.20 $ /month   5,882.93    
14150               3,674.00 - 4,690.00 $ /month   3,674.00    
14558               6,401.00 - 7,935.00 $ /month   6,401.00    
15189               1,878.00 - 1,878.00 $ /month   1,878.00    
16337               5,153.00 - 6,264.00 $ /month   5,153.00    
17609               4,900.00 - 6,125.00 $ /month   4,900.00    
18109              6,000.00 - 10,000.00 $ /month   6,000.00    
18281  5,000.00 - 15,000.00 $ /monthPlus Bonuses   5,000.00    
19265               3,293.00 - 4,542.00 $ /month   3,293.00    

                             to_salary yearly_hourly  
id                                                    
2662                20,971.00 $ /month         month  
2900                 3,865.66 $ /month         month  
3300                20,971.00 $ /month         month  
3317                 8,060.00 $ /month         month  
3345                 6,579.41 $ /month         month  
3622                 6,579.41 $ /month         month  
5949                 5,445.57 $ /month         month  
12333                3,500.00 $ /month         month  
13554                7,883.20 $ /month         month  
14150                4,690.00 $ /month         month  
14558                7,935.00 $ /month         month  
15189                1,878.00 $ /month         month  
16337                6,264.00 $ /month         month  
17609                6,125.00 $ /month         month  
18109               10,000.00 $ /month         month  
18281   15,000.00 $ /monthPlus Bonuses         month  
19265                4,542.00 $ /month         month

# View Up to...

In [67]:
filt = (pd.isna(df['salary']) == False) & (df['yearly_hourly'].str.contains('upto') == True)

df.loc[filt][['salary', 'from_salary', 'to_salary', 'yearly_hourly']]

salary from_salary to_salary yearly_hourly
id                                                        
42     Up to $32000.00         NaN  32000.00          upto
154    Up to $45000.00         NaN  45000.00          upto
2630      Up to $18.00         NaN     18.00      uptohour
2639      Up to $13.00         NaN     13.00      uptohour
2651      Up to $18.00         NaN     18.00      uptohour
...                ...         ...       ...           ...
14069     Up to $15.00         NaN     15.00      uptohour
20014     Up to $14.00         NaN     14.00      uptohour
21168     Up to $20.00         NaN     20.00      uptohour
21270  Up to $31500.00         NaN  31500.00          upto
21902  Up to $35000.00         NaN  35000.00          upto

[101 rows x 4 columns]

# View others

In [68]:
filt = (pd.isna(df['salary']) == False) & (pd.isna(df['yearly_hourly']) == True)

with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print(df.loc[filt][['salary', 'from_salary', 'to_salary', 'yearly_hourly']])

                                                  salary from_salary  \
id                                                                     
23                          Excellent Pay and Incentives         nan   
58                               Salary, plus commission         nan   
70                                      To be discussed.         nan   
92               bonus, 401K matching, medical, vacation         nan   
125                                                  DOE         nan   
179                       Negotiable based on experience         nan   
183                                    Competitive Wages         nan   
209    Burg Simpson offers excellent benefits and com...         nan   
225                  Excellent compensation and benefits         nan   
451                                        Yearly Salary         nan   
484                                    $300 Hiring Bonus         nan   
521                          Salary based on experience.        

I copied the data in "salary" to "salary_copy"

In [ ]:
df['salary_copy'] = df['salary']

In [ ]:
df[['salary', 'salary_copy']]

salary  \
id                                                         
0                                                    NaN   
1                                                    NaN   
2                                                    NaN   
3                                                    NaN   
4                                                    NaN   
...                                                  ...   
21995               120,000.00 - 160,000.00 $ /yearbonus   
21996                      45,000.00 - 60,000.00 $ /year   
21997                                                NaN   
21998                              25.00 - 28.00 $ /hour   
21999  40,000.00 - 46,000.00 $ /year+ annual bonus (u...   

                                             salary_copy  
id                                                        
0                                                    NaN  
1                                                    NaN  
2                                                    NaN  
3                                                    NaN  
4                                                    NaN  
...                                                  ...  
21995               120,000.00 - 160,000.00 $ /yearbonus  
21996                      45,000.00 - 60,000.00 $ /year  
21997                                                NaN  
21998                              25.00 - 28.00 $ /hour  
21999  40,000.00 - 46,000.00 $ /year+ annual bonus (u...  

[22000 rows x 2 columns]

Set "Yearly_Hourly" column to "hour" where "Salary" column contains "hour"

In [ ]:
filt_hour = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('hour') == True)

for i in df.index[filt_hour == True]:
    df.iloc[i, df.columns.get_loc('yearly_hourly')] = 'hour'

Now we got "hour" in these columns

In [ ]:
filt_hour = (pd.isna(df['salary']) == False) & (df['salary'].str.contains('hour') == True)

print(df.loc[filt_hour][['salary', 'yearly_hourly']])

                                    salary yearly_hourly
id                                                      
13                    9.00 - 13.00 $ /hour          hour
30                   62.00 - 81.00 $ /hour          hour
43                   15.00 - 16.00 $ /hour          hour
68     13.75 - 16.75 $ /hourYear End Bonus          hour
80                   40.00 - 50.00 $ /hour          hour
...                                    ...           ...
21909                17.00 - 21.00 $ /hour          hour
21912                20.00 - 22.00 $ /hour          hour
21913                50.00 - 55.00 $ /hour          hour
21927                30.00 - 35.00 $ /hour          hour
21998                25.00 - 28.00 $ /hour          hour

[1145 rows x 2 columns]


Take a look at the rows with no "hour" or "year" strings

In [ ]:
filt_salary_type = (pd.isna(df['yearly_hourly']) == True) & (pd.isna(df['salary']) == False)

rows_no_clasiff = df.loc[filt_salary_type]['salary']

with pd.option_context('display.max_rows', None,
                       'display.max_columns', None,
                       'display.precision', 3,
                       ):
    print('Rows with no year/hour classification: ', rows_no_clasiff.count())
    print(rows_no_clasiff)



Rows with no year/hour classification:  545
id
23                            Excellent Pay and Incentives
42                                         Up to $32000.00
58                                 Salary, plus commission
70                                        To be discussed.
92                 bonus, 401K matching, medical, vacation
125                                                    DOE
154                                        Up to $45000.00
179                         Negotiable based on experience
183                                      Competitive Wages
209      Burg Simpson offers excellent benefits and com...
225                    Excellent compensation and benefits
451                                          Yearly Salary
484                                      $300 Hiring Bonus
521                            Salary based on experience.
659                                    Bonus Opportunities
897                   Competitive Base Salary + Commission
1056     

We got some values like the following:\
-> Excellent Pay and Incentives\
-> Up to 32000.00\
-> 108,887.00 - 141,555.00\
-> Up to 18.00\
-> 17,688.94 - 20,971.00  /month

In [ ]:
numbers = [i for i in range(10)]

num_strings = [str(x) for x in numbers]

#------------------------------------------------------------------------------------------------

filt_salary_type = (pd.isna(df['yearly_hourly']) == True) & (pd.isna(df['salary']) == False)

rows_no_clasiff = df.loc[filt_salary_type]['salary']

#------------------------------------------------------------------------------------------------

for i in df.index[filt_salary_type == True]:
    salary_cell = df.loc[i, 'salary']
    if any(num in salary_cell for num in num_strings):
        print(i, salary_cell)

42 Up to $32000.00
92 bonus, 401K matching, medical, vacation
154 Up to $45000.00
484 $300 Hiring Bonus
1067 $4,000 Sign-On Bonus
1103 $1,100.00+ /week
1179 Employee Health/Dental Benefits; 401k Employer Match
2086 401K, Medical, Dental
2424 108,887.00 - 141,555.00
2630 Up to $18.00
2639 Up to $13.00
2651 Up to $18.00
2655 Up to $14.00
2662 17,688.94 - 20,971.00 $ /month
2669 Up to $18.00
2717 Up to $15.00
2792 Up to $30000.00
2851 Up to $10.00
2900 3,137.89 - 3,865.66 $ /month
3059 Up to $15.00
3062 Bonus and full benefits package (401K, etc.)
3087 Up to $13.50
3102 Up to $31000.00
3150 Up to $14.00
3165 Up to $150000.00
3251 Up to $92000.00
3300 17,688.94 - 20,971.00 $ /month
3317 6,833.33 - 8,060.00 $ /month
3345 4,224.31 - 6,579.41 $ /month
3446 Up to $40.00
3600 Up to $14.00
3622 4,905.22 - 6,579.41 $ /month
3674 Up to $55.00
3711 Up to $92000.00
3715 Up to $18.00
3735 Up to $42.00
3834 Up to $25.00
4251 Up to $60000.00
4789 Salary, Commissions, Bonuses, Health Insurance, 401K wit

In [ ]:
df[['salary']]

salary
id                                                      
0                                                    NaN
1                                                    NaN
2                                                    NaN
3                                                    NaN
4                                                    NaN
...                                                  ...
21995               120,000.00 - 160,000.00 $ /yearbonus
21996                      45,000.00 - 60,000.00 $ /year
21997                                                NaN
21998                              25.00 - 28.00 $ /hour
21999  40,000.00 - 46,000.00 $ /year+ annual bonus (u...

[22000 rows x 1 columns]

In [ ]:
df

country country_code date_added has_expired  \
0      United States of America           US        NaN          No   
1      United States of America           US        NaN          No   
2      United States of America           US        NaN          No   
3      United States of America           US        NaN          No   
4      United States of America           US        NaN          No   
...                         ...          ...        ...         ...   
21995  United States of America           US        NaN          No   
21996  United States of America           US        NaN          No   
21997  United States of America           US        NaN          No   
21998  United States of America           US        NaN          No   
21999  United States of America           US        NaN          No   

              job_board                                    job_description  \
0      jobs.monster.com  TeamSoft is seeing an IT Support Specialist to...   
1      jobs.monster.com  The Wisconsin State Journal is seeking a flexi...   
2      jobs.monster.com  Report this job About the Job DePuy Synthes Co...   
3      jobs.monster.com  Why Join Altec? If you’re considering a career...   
4      jobs.monster.com  Position ID#  76162 # Positions  1 State  CT C...   
...                 ...                                                ...   
21995  jobs.monster.com  This is a major premier Cincinnati based finan...   
21996  jobs.monster.com  Luxury homebuilder in Cincinnati seeking multi...   
21997  jobs.monster.com  RE: Adobe AEM-          Client -          Loca...   
21998  jobs.monster.com  Jernberg Industries was established in 1937 an...   
21999  jobs.monster.com  Contract AdministratorCan you be the point per...   

                                               job_title             job_type  \
0                   IT Support Technician Job in Madison   Full Time Employee   
1                Business Reporter/Editor Job in Madison            Full Time   
2      Johnson & Johnson Family of Companies Job Appl...  Full Time, Employee   
3                        Engineer - Quality Job in Dixon            Full Time   
4           Shift Supervisor - Part-Time Job in Camphill   Full Time Employee   
...                                                  ...                  ...   
21995  Assistant Vice President - Controller Job in C...            Full Time   
21996                       Accountant Job in Cincinnati            Full Time   
21997                    AEM/CQ developer Job in Chicago            Full Time   
21998  Electrician - Experienced Forging Electrician ...   Full Time Employee   
21999           Contract Administrator Job in Cincinnati            Full Time   

                                                location  \
0                                      Madison, WI 53702   
1                                      Madison, WI 53708   
2      DePuy Synthes Companies is a member of Johnson...   
3                                              Dixon, CA   
4                                           Camphill, PA   
...                                                  ...   
21995                                     Cincinnati, OH   
21996                               Cincinnati, OH 45236   
21997                                  Chicago, IL 60602   
21998                                  Chicago, IL 60609   
21999                                     Cincinnati, OH   

                                         organization  \
0                                                 NaN   
1                             Printing and Publishing   
2                     Personal and Household Services   
3                                    Altec Industries   
4                                              Retail   
...                                               ...   
21995                                             NaN   
21996  Construction - Residential & Commercial/Office   
21997                      

Dictionary + astype to convert column dtype

In [ ]:
# convert_dict = {'salary': float}

# df = df.astype(convert_dict)

In [ ]:
df.dtypes

country            object
country_code       object
date_added         object
has_expired        object
job_board          object
job_description    object
job_title          object
job_type           object
location           object
organization       object
page_url           object
salary             object
sector             object
uniq_id            object
dtype: object

'Apply' method to convert column dtype

Empty cells by column
What can we do about these columns?

-> date_added:
 <br />-> job_type: 
 <br />-> organization: 
 <br />-> salary:
 <br />-> sector: 

In [ ]:
df.isnull().sum()

country                0
country_code           0
date_added         21878
has_expired            0
job_board              0
job_description        0
job_title              0
job_type            1628
location               0
organization        6867
page_url               0
salary             18554
sector              5194
uniq_id                0
dtype: int64